src/node_file.cc

```diff
@@ -2480,6 +2480,64 @@ static void WriteString(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+static void WriteFileUtf8(const FunctionCallbackInfo<Value>& args) {
+  // Fast C++ path for fs.writeFileSync(path, data) with utf8 encoding
+  // (file, data, options.flag, options.mode)
+
+  Environment* env = Environment::GetCurrent(args);
+  auto isolate = env->isolate();
+
+  CHECK_GE(args.Length(), 4);
+
+  CHECK(args[1]->IsString());
+
+  CHECK(args[2]->IsInt32());
+  const int flags = args[2].As<Int32>()->Value();
+
+  CHECK(args[3]->IsInt32());
+  const int mode = args[3].As<Int32>()->Value();
+
+  uv_file file;
+  uv_fs_t req;
+
+  bool is_fd = args[0]->IsInt32();
+
+  // Check for file descriptor
+  if (is_fd) {
+    file = args[0].As<Int32>()->Value();
+  } else {
+    BufferValue path(env->isolate(), args[0]);
+    CHECK_NOT_NULL(*path);
+    if (CheckOpenPermissions(env, path, flags).IsNothing()) return;
+
+    FS_SYNC_TRACE_BEGIN(open);
+    file = uv_fs_open(nullptr, &req, *path, flags, mode, nullptr);
+    FS_SYNC_TRACE_END(open);
+
+    if (req.result < 0) {
+      uv_fs_req_cleanup(&req);
+      return env->ThrowUVException(req.result, "open", nullptr, path.out());
+    }
+  }
+
+  node::Utf8Value value(isolate, args[1]);
+  uv_buf_t uvbuf = uv_buf_init(*value, value.length());
+
+  FS_SYNC_TRACE_BEGIN(write);
+  int err = uv_fs_write(nullptr, &req, file, &uvbuf, 1, 0, nullptr);
+  FS_SYNC_TRACE_END(write);
+
+  if (!is_fd) {
+    FS_SYNC_TRACE_BEGIN(close);
+    CHECK_EQ(0, uv_fs_close(nullptr, &req, file, nullptr));
+    FS_SYNC_TRACE_END(close);
+  }
+  uv_fs_req_cleanup(&req);
+
+  if (err < 0) {
+    return env->ThrowUVException(err, "write");
+  }
```

Comment:
Shouldn't we return the bytes written?
```suggestion
  if (err < 0) {
    return env->ThrowUVException(err, "write");
  }
  
  args.GetReturnValue().Set(err);
```

Reply:
No, `writeFileSync` doesn't return anything.

Reply:
According to documentation `fs.writeFileSync` returns `undefined`. Ref: https://nodejs.org/dist/latest-v20.x/docs/api/fs.html#fswritefilesyncfile-data-options

Reply:
Yes, I understand that but nothing guarantees that writing to the file is complete in one syscall (libuv does that for you but one of the partial writes might fail) and you need to check that by using the return value of `uv_fs_write()` as I think it's done at the JS level https://github.com/nodejs/node/blob/main/lib/fs.js#L2296-L2300.

Reply:
Fwiw, I spoke to some people and it didn't  sound necessary in this case as it is just a single utf8 string instead of buffers, but not sure.

Reply:
@santigimeno I think you're confusing `writeSync` with `writeFileSync`

Reply:
I'd rather guess not what the amount of data the os decides to write on every `write()` call.
Anyway, here's an example that demonstrates the issue:
```bash
// Create and mount  a fs of 500K
$ dd if=/dev/zero of=filesystem.img bs=1K count=500
$ sudo mkfs.ext2 filesystem.img
$ mkdir mountpoint
$ sudo mount -o loop filesystem.img mountpoint
```
```js
// Running this with sudo inside mountpoint.
// Write 16M string to mountpoint/hello.txt
const fs = require('fs');
const str = 'a'.repeat(16e6);
fs.writeFileSync('./hello.txt', str);
```
Using `v20.6.1` the output, as expected, is:
```
node:internal/fs/utils:352
    throw err;
    ^

Error: ENOSPC: no space left on device, write
    at Object.writeSync (node:fs:951:3)
    at Object.writeFileSync (node:fs:2344:26)
    at Object.<anonymous> (/tmp/mountpoint/test.js:4:4)
    at Module._compile (node:internal/modules/cjs/loader:1241:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1295:10)
    at Module.load (node:internal/modules/cjs/loader:1091:32)
    at Module._load (node:internal/modules/cjs/loader:938:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:83:12)
    at node:internal/main/run_main_module:23:47 {
  errno: -28,
  syscall: 'write',
  code: 'ENOSPC'
}
```
whereas with this patch it doesn't fail, though by inspecting the syscalls via `strace` the partial write with a `ENOSPC` failure can be observed:
```
664365 pwrite64(23, "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"..., 16000000, 0) = 442368
664365 pwrite64(23, "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"..., 15557632, 442368) = -1 ENOSPC (No space left on device)
```

Reply:
Just to check, are you using the latest commit of this patch? We could possibly fix this by verifying the amount of bytes written matches expected value from the string?

Reply:
Yes, same issue. The problem is that, in order to catch this problem you have to call again to `uv_fs_write()` because `uv_fs_write()` won't notify any error if at least some data has been successfully written. That's the reason you need to return the bytes written and compare to the bytes in the string. (Probably you can do everything from C++ instead of going back to JS)

Reply:
I'll add a check at the end in C++ to ensure number of bytes written returned by write() matches expectations from data string, and throw if not.

Reply:
Disregard last, rewriting to use while like original JS version.

Reply:
No, no observable changes in behavior. Retry with the remainder of the buffer like the JS code does.

Reply:
Rewrote to use a write loop like JS, should have the same behavior now.